odoo/addons/base/models/qweb.py

```diff
@@ -251,115 +109,120 @@ def _render(self, template, values=None, **options):
         :param template: template identifier
         :param dict values: template values to be used for rendering
         :param options: used to compile the template (the dict available for the rendering is frozen)
-            * ``load`` (function) overrides the load method
-            * ``profile`` (float) profile the rendering (use astor lib) (filter
-              profile line with time ms >= profile)
+            * ``load`` (function) overrides the load method (returns: (template, ref))
+            * ``profile`` (boolean) profile the rendering
         """
-        values = values or {}
-        body = []
-        self.compile(template, options)(self, body.append, values)
-        joined = u''.join(body)
-        if not values.get('__keep_empty_lines'):
-            joined = QWeb._empty_line.sub('\n', joined.strip())
-        return joined.encode('utf8')
+        render_template = self._compile(template, options)
+        return render_template(self, values or {})
 
-    def compile(self, template, options):
-        """ Compile the given template into a rendering function::
+    def _compile(self, template, options):
+        """ Compile the given template into a rendering function (generator)::
 
-            render(qweb, append, values)
+            render(qweb, values)
 
-        where ``qweb`` is a QWeb instance, ``append`` is a unary function to
-        collect strings into a result, and ``values`` are the values to render.
+        where ``qweb`` is a QWeb instance and ``values`` are the values to render.
         """
         if options is None:
             options = {}
 
+        element, document, ref = self._get_template(template, options)
+        if not ref:
+            ref = element.get('t-name', str(document))
+
+        options['ref'] = ref
+        if 'profile' in options:
+            options['document'] = document if isinstance(document, str) else str(document, 'utf-8')
+
         _options = dict(options)
         options = frozendict(options)
 
-        element, document = self.get_template(template, options)
-        name = element.get('t-name', 'unknown')
-
         _options['template'] = template
-        _options['ast_calls'] = []
         _options['root'] = element.getroottree()
         _options['last_path_node'] = None
         if not options.get('nsmap'):
             _options['nsmap'] = {}
 
-        # generate ast
+        def manage_exception(e, path, code=None):
```

Comment:
It does not format html. With error it displays the xml line corresponding to the error. This is the part sent to the client to allow him to fix his template.

Reply:
> It does not format html. With error it displays the xml line corresponding to the error.

Same diff, point is that formatting the XML element to it can be included as part of the error message is the only thing the helper really does, so it would be much simpler if it did *just* that and was called and the rest (including actually creating and raising the QWebException) was done normally by the caller.